TFEngine.py

```diff
@@ -1857,86 +1857,118 @@ def search(self, dataset, do_eval=True, output_layer_name="output", output_file=
+
+        assert n_batch * (out_beam_sizes[t] or 1) == len(outputs[t])
+        if targets[t] is not None:
+          assert n_batch == len(targets[t])
+
+        if output_layers[t].output.dim == 256 and output_layers[t].output.sparse:
+          # Interpret output as bytes/utf8-string.
+          outputs[t] = bytearray(outputs[t]).decode("utf8")
+
       for i in range(len(seq_idx)):
-        if out_beam_size is None:
-          print("seq_idx: %i, seq_tag: %r, output: %r" % (seq_idx[i], seq_tag[i], output[i]), file=log.v4)
-          out_idx = i
-        else:
-          print("seq_idx: %i, seq_tag: %r, outputs: %r" % (
-            seq_idx[i], seq_tag[i], output[i * out_beam_size:(i + 1)*out_beam_size]), file=log.v4)
-          out_idx = i * out_beam_size
-        if target_key and dataset.can_serialize_data(target_key):
-          print("  ref:", dataset.serialize_data(key=target_key, data=targets[i]), file=log.v4)
-          if out_beam_size is None:
-            print("  hyp:", dataset.serialize_data(key=target_key, data=output[out_idx]), file=log.v4)
+        for t in range(num_targets):
+          if out_beam_sizes[t] is None:
+            print("seq_idx: %i, seq_tag: %r, output: %r" % (seq_idx[i], seq_tag[i], outputs[t][i]), file=log.v4)
+            out_idx = i
           else:
-            assert beam_scores is not None
-            for b in range(out_beam_size):
-              print(
-                "  hyp %i, score %f:" % (b, beam_scores[i][b]),
-                dataset.serialize_data(key=target_key, data=output[out_idx + b]),
-                file=log.v4)
+            print("seq_idx: %i, seq_tag: %r, outputs: %r" % (
+              seq_idx[i], seq_tag[i], outputs[t][i * out_beam_sizes[t]:(i + 1)*out_beam_sizes[t]]), file=log.v4)
+            out_idx = i * out_beam_sizes[t]
+          if target_keys[t] and dataset.can_serialize_data(target_keys[t]):
+            print("  ref:", dataset.serialize_data(key=target_keys[t], data=targets[t][i]), file=log.v4)
+            if out_beam_sizes[t] is None:
+              print("  hyp:", dataset.serialize_data(key=target_keys[t], data=outputs[t][out_idx]), file=log.v4)
+            else:
+              assert beam_scores[t] is not None
+              for b in range(out_beam_sizes[t]):
+                print(
+                  "  hyp %i, score %f:" % (b, beam_scores[t][i][b]),
+                  dataset.serialize_data(key=target_keys[t], data=outputs[t][out_idx + b]),
+                  file=log.v4)
+
         if out_cache is not None:
           corpus_seq_idx = dataset.get_corpus_seq_idx(seq_idx[i])
           assert corpus_seq_idx not in out_cache
           seq_idx_to_tag[corpus_seq_idx] = seq_tag[i]
           if out_beam_size is None:
-            out_cache[corpus_seq_idx] = dataset.serialize_data(key=target_key, data=output[out_idx])
+            out_cache[corpus_seq_idx] = '\t'.join(dataset.serialize_data(key=target_keys[t],
```

Comment:
I simply used tabs to separate the outputs. Maybe different files would be better?

Reply:
Yes, no tabs.
But I think separate output files might make it more complicated than necessary. Why not just a dict? (But only for the case `isinstance(output_layer_names, list)`.)